Automatic compaction by coordinators

jihoonson opened the pull request:
Part of https://github.com/druid-io/druid/issues/4479.

This patch introduces the automatic compaction by coordinators.

## How it works

Each run, the Druid coordinator compacts small segments abutting each other. It first finds the segments to compact together based on the _segment search policy_. Once it finds some segments, it launches a compactTask to compact those segments. The maximum number of running compact tasks is limited by `max(sum of worker capacity * compactionTaskSlotRatio, 1)`.

### Segment Search Policy

Currently there is only one segment search policy, newest segment first policy.

#### Newest Segment First Policy

This policy searches the segments of _all dataSources_ in inverse order of their intervals.
For example, let me assume there are 3 dataSources (`ds1`, `ds2`, `ds3`) and 5 segments (`seg_ds1_2017-10-01_2017-10-02`, `seg_ds1_2017-11-01_2017-11-02`, `seg_ds2_2017-08-01_2017-08-02`, `seg_ds3_2017-07-01_2017-07-02`, `seg_ds3_2017-12-01_2017-12-02`) for those dataSources. The segment name indicates its dataSource and interval. The search result of newestSegmentFirstPolicy is [`seg_ds3_2017-12-01_2017-12-02`, `seg_ds1_2017-11-01_2017-11-02`, `seg_ds1_2017-10-01_2017-10-02`, `seg_ds2_2017-08-01_2017-08-02`, `seg_ds3_2017-07-01_2017-07-02`]. 

Every run, this policy starts searching segments for the (very latest interval - skipOffsetFromLatest). This is to handle the late segments ingested to realtime dataSources.

This policy currenlty cannot handle the situation when a shard consists of a lot of small segments, thereby its total size exceeds the targetCompactionSizebytes. If it finds such shards, it simply skips them.

I did some benchmark for the performance of this policy. It took about 1 seconds to iterate 10000000 segments in my mac. I think it's fine because 1) having such many segments is not common and 2) the term between each run of this policy is usually much larger than 1 seconds (it's 30 mins by default).

## How you can test it

You can post, get, and delete a compaction config for a dataSource with a new coordinator API ```http://{COORDINATOR_IP:PORT}/druid/coordinator/v1/config/{dataSource}```.

A sample compaction config is:

```json
{
  "dataSource": "wikiticker",
  "targetCompactionSizeBytes": 800000000,
  "skipOffsetFromLatest": "P1D"
}
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/5102)
<!-- Reviewable:end -->

jihoonson commented:
This patch is ready for review.

himanshug commented:
started reviewing this PR, some explanatory javadocs on `NewestSegmentFirstIterator` class would be helpful.

himanshug commented on server/src/main/java/io/druid/server/coordinator/DruidCoordinatorRuntimeParams.java:
```diff
@@ -303,9 +304,30 @@ public Builder withSegmentReplicantLookup(SegmentReplicantLookup lookup)
+                )
+            );
+          }
+      );
```
is this always done or only when compaction is enabled and this information is needed ?

jihoonson replied:
This is always done. Currently, all segments of the all dataSources are being gathered to emit some statistics in ```DruidCoordinatorLogger```. I only changed the type of ```dataSources``` in ```DruidCoordinatorRuntimeParams``` from ```Set<ImmutableDataSource>``` to ```Map<String, VersionedIntervalTimeline<String, DataSegment>>``` to use this information in ```DruidCoordinatorSegmentCompactor``` as well. 

Another alternative is to keep the current type of ```dataSources``` in ```DruidCoordinatorRuntimeParams``` and convert it to ```Map<String, VersionedIntervalTimeline<String, DataSegment>>``` in ```DruidCoordinatorSegmentCompactor``` when it's needed, but this may increase the memory usage. As an extreme example, there might be a lot of segments and the compaction might be enabled for all dataSources. In that case, this alternative approach will double the memory consumption. I expect users to enable the automatic compaction for all their realtime dataSources as well as some batch dataSources, so the memory burden might not be small.

himanshug replied:
its alright given specially since it was there already.

himanshug commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,183 @@
+
+  private final CompactionSegmentSearchPolicy policy = new NewestSegmentFirstPolicy();
+  private final IndexingServiceClient indexingServiceClient;
+  private final Set<String> queryIds = new HashSet<>();
```
I don't see this getting initialized anywhere when coordinator starts .. e.g. on a leadership change , new coordinator would not take compaction tasks into account which were started by previous leader.

himanshug replied:
I think the word query is used because of the current naming in `IndexingServiceClient` , it might still make sense to use word `task` instead in various logs and variables in this class e.g. instead of queryIds using compationTaskIds .

jihoonson replied:
> I don't see this getting initialized anywhere when coordinator starts .. e.g. on a leadership change , new coordinator would not take compaction tasks into account which were started by previous leader.

Good point! I'm thinking to get the number of running compaction tasks when ```DruidCoordinatorSegmentCompactor.run()``` is called, and I'll add it soon.

BTW, the refactoring done in https://github.com/druid-io/druid/pull/5149 will be helpful to implement it. Some classes are added or moved to ```druid-api``` to be reused by overlord clients without code duplication. It will be easy to get the number of running compaction tasks this once https://github.com/druid-io/druid/pull/5149 is merged.

jihoonson replied:
> I think the word query is used because of the current naming in IndexingServiceClient , it might still make sense to use word task instead in various logs and variables in this class e.g. instead of queryIds using compationTaskIds .

Ok. I'll rename it together.

himanshug commented on docs/content/design/coordinator.md:
```diff
@@ -241,6 +276,11 @@ Returns total size and count for each interval within given isointerval.
 
+#### Compaction Configs
+
+* `/druid/coordinator/v1/config/{dataSource}`
```
is this endpoint specific to compaction config or is it meant to support any dataSource level dynamic config?

jihoonson replied:
It's specific to compactionConfig. I changed the endpoint to ```/druid/coordinator/v1/config/compaction/{dataSource}```.

himanshug commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,183 @@
+
+  private static final Logger LOG = new Logger(DruidCoordinatorSegmentCompactor.class);
+
+  private final CompactionSegmentSearchPolicy policy = new NewestSegmentFirstPolicy();
```
is it intentional that the policy is not plug-able ?

jihoonson replied:
Yes. I plan to make another policy considering the dataSource priority. I'll make this plugable in a follow-up PR.

himanshug commented:
is this feature general enough that any druid user can just turn it on and forget about it? if not, then can you please add some blurb to the docs explaining when it should be enabled and when not.

jihoonson commented:
@himanshug thank you for the review. I added more documents to explain when the automatic compaction is useful (https://github.com/druid-io/druid/pull/5102/files#diff-fd047eb9f74cdf2a732aa0d3f72772feR341, https://github.com/druid-io/druid/pull/5102/files#diff-745f7bb3fc90d4a31c1658733f19159cR339, https://github.com/druid-io/druid/pull/5102/files#diff-507910e969a3afeceb1f9dbe9be0a644R138) and added javadocs to NewestSegmentFirstIterator.

himanshug commented:
@jihoonson please let me know when you're done with this, I'm assuming https://github.com/druid-io/druid/pull/5149 needs to be merged first to complete this PR.

jihoonson commented:
@himanshug thanks. I'll let you know when it's ready.

jihoonson commented:
@himanshug addressed your comments. It's ready for review now.

jon-wei commented on server/src/main/java/io/druid/client/indexing/ClientCompactQuery.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.List;
+import java.util.Map;
+
+public class ClientCompactQuery
```
hm, I feel like is more like a "Specification" than a "Query", maybe rename this to ClientCompactSpec

jihoonson replied:
The name of the taskSpecs used by overlords have a pattern of ```Client***Query```. I guess it's because the tasks submitted by coordinators can be thought of as queries to overlords. I'm not sure this is a good convention because it makes me somewhat confused at first.

I think it's better to move all task classes to some module (maybe  ```druid-api```) which allows all other modules and user libraries to depend on. This will let us to share the same class in all places. If we decide to go for it, I would like to do in a follow-up PR. What do you think?

gianm replied:
@jihoonson If you mean the actual "CompactionTask" etc classes, I think probably moving something so heavy from druid-indexing-service all the way down to druid-api would probably require collapsing a ton of druid modules into one giant druid-core module. I guess we could do that but it seems like a big change. Do you think it's worth it?

gianm replied:
I am definitely open to an argument that collapsing some modules is a good idea because it solves problems like this, and because the multi-module design of Druid didn't accomplish what it originally set out to accomplish (making it easy for people to write extensions or clients without depending on a ton of code). It's still required for many extensions, and basically all clients, to at least depend on druid-processing if not druid-server too.

If this is an argument that you want to make, could you please propose a list of modules to collapse and some comments about why?

jihoonson replied:
> I think probably moving something so heavy from druid-indexing-service all the way down to druid-api would probably require collapsing a ton of druid modules into one giant druid-core module.

@gianm would you elaborate on this? I simply thought some classes used as the input or return types for user APIs (like ```CompactionTask``` or ```TaskStatus```) can be moved to ```druid-api``` rather than collapsing core modules to a single giant one.

gianm replied:
@jihoonson I meant that the the API classes like `CompactionTask` are very heavy with implementation, and likely have enough transitive class dependencies that it is impractical to move them to `druid-api` without pulling along a lot of other classes too. However, like I said I am open to an argument that this is worth doing, since I'm not sure the large number of core druid modules is buying us much vs. a small number of larger modules.

himanshug replied:
maybe in the far future... we can probably start by bringing druid-api, druid-commons, druid-processing, java-utils in a druid-core module as pretty much all extensions depend on them anyway.
later we can possibly move some core stuff from druid-server and druid-indexing-service to core as well.

himanshug replied:
also, I do agree with @jon-wei on naming it *spec instead of *query even if that does not match with existing not-so-great convention . we don't need to change others but this class could have the saner name . "ClientCompactQuery" gave me the perception that it was a new query type.

himanshug replied:
nvrmnd, as I look at more code .. I think its very unintuitive and probably should be fixed in a separate PR. to me, there are two weird things about these classes.
- xxxQuery gives the impression that they are some query types as in groupBy
- xxxQuery serialized json is expected to be deserializable into a Task object. so, probably we should use corresponding Task classes themselves rather that creating these new xxxQuery (which are tasks really)

himanshug replied:
https://github.com/druid-io/druid/issues/5222

jihoonson replied:
@himanshug thanks for raising #5222. I think the best solution is to use the same class in ```druid-server``` and ```druid-indexing-service``` modules. To do so, some classes should be moved to a common module which both ```druid-server``` and ```druid-indexing-service``` modules depend on. 

@gianm thanks for the explanation. Yes, there are a lot of class dependencies on indexing task implementations, so collapsing several druid modules might be better. I will raise an issue to discuss it.

jihoonson commented on docs/content/configuration/coordinator.md:
```diff
@@ -106,6 +106,8 @@ Issuing a GET request at the same URL will return the spec that is currently in
 |`killDataSourceWhitelist`|List of dataSources for which kill tasks are sent if property `druid.coordinator.kill.on` is true.|none|
 |`killAllDataSources`|Send kill tasks for ALL dataSources if property `druid.coordinator.kill.on` is true. If this is set to true then `killDataSourceWhitelist` must not be specified or be empty list.|false|
 |`maxSegmentsInNodeLoadingQueue`|The maximum number of segments that could be queued for loading to any given server. This parameter could be used to speed up segments loading process, especially if there are "slow" nodes in the cluster (with low loading speed) or if too much segments scheduled to be replicated to some particular node (faster loading could be preferred to better segments distribu ...
+|`compactionConfigs`|Compction config list. See the below [Compaction Config](#compaction-config)|none|
```
Thanks. Fixed.

jihoonson commented on docs/content/configuration/coordinator.md:
```diff
@@ -106,6 +106,8 @@ Issuing a GET request at the same URL will return the spec that is currently in
 |`killAllDataSources`|Send kill tasks for ALL dataSources if property `druid.coordinator.kill.on` is true. If this is set to true then `killDataSourceWhitelist` must not be specified or be empty list.|false|
 |`maxSegmentsInNodeLoadingQueue`|The maximum number of segments that could be queued for loading to any given server. This parameter could be used to speed up segments loading process, especially if there are "slow" nodes in the cluster (with low loading speed) or if too much segments scheduled to be replicated to some particular node (faster loading could be preferred to better segments distribu ...
+|`compactionConfigs`|Compction config list. See the below [Compaction Config](#compaction-config)|none|
+|`compactionTaskSlotRatio`|The ratio of the total task slots to the copmaction task slots|0.1|
```
Good point. I added a new ```maxCompactionTaskSlots``` configuration.

jihoonson commented on docs/content/design/coordinator.md:
```diff
@@ -38,6 +38,44 @@ Balancing Segment Load
+
+Each run, the Druid coordinator compacts small segments abutting each other. This is useful when you have a lot of small
+segments which may degrade the query performance as well as increasing the disk usage. Note that the data for an interval
+cannot be rolled up across the partitions.
```
Sounds good. Changed.

jihoonson commented on docs/content/design/coordinator.md:
```diff
@@ -38,6 +38,44 @@ Balancing Segment Load
+Compact tasks might fail due to some reasons.
+
+- If the input segments of a compact task are removed or overshadowed before it starts, that compact task fails immediately.
+- If a task of a higher priority acquires a lock for an interval overlapping with the interval of a compact task, the compat task fails.
```
Thanks. Fixed.

jihoonson commented on docs/content/design/coordinator.md:
```diff
@@ -38,6 +38,44 @@ Balancing Segment Load
+This is to handle the late segments ingested to realtime dataSources.
+
+<div class="note caution">
+This policy currenlty cannot handle the situation when a shard consists of a lot of small segments, thereby its total size exceeds the <a href="../configuration/coordinator.html#compaction-config">targetCompactionSizebytes</a>.
```
Thanks. Fixed.

jihoonson commented on indexing-service/src/main/java/io/druid/indexing/common/task/CompactionTask.java:
```diff
@@ -475,12 +475,21 @@ private static DimensionSchema createDimensionSchema(
+                                                                .collect(Collectors.toList());
+          throw new ISE(
+              "Specified segments in the spec are different from the current used segments. "
+              + "There are unused segments[%s] and missing segments[%s] in the spec.",
```
It makes more sense to me because ```unknownSegments``` can include nonexistent segments as well as unused segments. Changed.

jon-wei commented:
@himanshug Did you have any more comments on this?

jihoonson commented:
@jon-wei thanks for the review. @himanshug do you have more comments?

himanshug commented:
@jihoonson will review it again today. thanks.

jihoonson commented:
@himanshug thanks.

himanshug commented on docs/content/design/coordinator.md:
```diff
@@ -38,6 +38,44 @@ Balancing Segment Load
+This is to handle the late segments ingested to realtime dataSources.
+
+<div class="note caution">
+This policy currently cannot handle the situation when a shard consists of a lot of small segments, thereby its total size exceeds the <a href="../configuration/coordinator.html#compaction-config">targetCompactionSizebytes</a>.
```
what does "shard" refer to ? ... a partition-set interval ?

jihoonson replied:
Yes.. The above sentence looks not clear, so I changed to ```This policy currently cannot handle the situation when there are a lot of small segments which have the same interval,
and their total size exceeds targetCompactionSizebytes.```.

himanshug commented on docs/content/design/coordinator.md:
```diff
@@ -38,6 +38,44 @@ Balancing Segment Load
+
+Each run, the Druid coordinator compacts small segments abutting each other. This is useful when you have a lot of small
+segments which may degrade the query performance as well as increasing the disk usage. Note that the data for an interval
+cannot be compacted across the partitions.
```
partition == partition set interval ?

jihoonson replied:
Changed to ```across the partitions```.

himanshug commented on docs/content/development/extensions-core/kafka-ingestion.md:
```diff
@@ -339,12 +339,10 @@ keeping the number of segments low are:
-There is ongoing work to support automatic segment compaction of sharded segments as well as compaction not requiring
-Hadoop (see [here](https://github.com/druid-io/druid/pull/1998) and [here](https://github.com/druid-io/druid/pull/3611)
-for related PRs).
+Currently, the recommended way of doing this is turning on the [automatic compaction of coordinators](../../design/coordinator.html#compacting-segments).
```
not sure if we should recommend one over the other at this time when new feature hasn't been used much and comes with cautionary notes. I think we should simply describe both options and let users choose whatever is convenient to them.
so far, I like batch re-indexing as it is relatively simpler to understand for users already using batch indexing and works for all cases.

makes sense ?

jihoonson replied:
Sounds make sense. Changed as suggested.

himanshug commented on docs/content/ingestion/stream-pull.md:
```diff
@@ -335,3 +335,18 @@ This can be difficult to manage at scale, especially with multiple partitions.
+## Deployment Notes
+
+Stream ingestion may generate a large number of small segments because it's difficult to optimize the segment size at
+ingestion time. The number of segments will increase over time, and this might cuase the query performance issue. 
```
s/cuase/cause

jihoonson replied:
Thanks. Fixed.

himanshug commented on docs/content/development/extensions-core/kafka-ingestion.md:
```diff
@@ -339,12 +339,10 @@ keeping the number of segments low are:
 In many production installations which have been ingesting events for a long period of time, these suggestions alone
```
not so much related to this PR, but the section "On the Subject of Segments" looks completely out-of-date and gives false idea of KIS always generating a lot of small segments and compaction being a mandatory requirement. would you be able to update this ?

jihoonson replied:
Do you mean we can avoid the small segments problem of KIS by setting a longer taskDuration and enabling incremental handoffs?

himanshug replied:
segment size is not so much impacted by taskDuration but by "taskCount" and "maxRowsInSegment" settings. After https://github.com/druid-io/druid/pull/4815 , we don't create one segment per kafka partition so, in general, segments generated shouldn't necessarily be small anymore.

@pjain1 can you please do a separate PR to fix the doc based on updates in #4815 ?

himanshug commented on docs/content/ingestion/stream-pull.md:
```diff
@@ -335,3 +335,18 @@ This can be difficult to manage at scale, especially with multiple partitions.
+Stream ingestion may generate a large number of small segments because it's difficult to optimize the segment size at
+ingestion time. The number of segments will increase over time, and this might cuase the query performance issue. 
+
+To mitigate this problem, it is recommended that scheduled re-indexing
```
this blurb appears to be repeated in multiple places.
how about creating a separate page that talks about segment compaction options available in Druid and linking to that page from all different pages?

jihoonson replied:
I guess you meant the paragraph about the needs for reindexing for steaming ingestion is repeated in several places because new sections for describing segment compaction are already added (https://github.com/druid-io/druid/pull/5102/files#diff-1401189ed202eac3beb0b4e0b4d535cbR41, https://github.com/druid-io/druid/pull/5102/files#diff-ad94e83d5b5a9cf743a09e7ccb0d9a95R127).

But, this looks a quite short paragraph (like only 6 lines) to create a new separate page. Do you think it's worthwhile?

himanshug replied:
I was writing same comments in multiple places and thought maybe a separate page just talking about how to maintain segments of optimal size and what causes small segments etc etc could be created. i don't have any super strong feelings about it.

jihoonson replied:
Sounds good. I added a new document.

himanshug commented on indexing-service/src/main/java/io/druid/indexing/overlord/hrtr/HttpRemoteTaskRunner.java:
```diff
@@ -1335,11 +1336,11 @@ void taskAddedOrUpdated(final TaskAnnouncement announcement, final WorkerHolder
     )
     {
-      super(taskId, worker, location);
+      super(taskId, task == null ? null : task.getType(), worker, location);
```
taskType should be updated in setTask(Task) method as well.

jihoonson replied:
Thanks. Added.

himanshug commented on indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java:
```diff
@@ -473,15 +500,28 @@ public Response getPendingTasks(@Context final HttpServletRequest req)
   @Produces(MediaType.APPLICATION_JSON)
-  public Response getRunningTasks(@Context final HttpServletRequest req)
+  public Response getRunningTasks(
+      @QueryParam("type") String taskType,
```
so all the taskType business is added to support this use case ? why did we need this ?

jihoonson replied:
First, in this PR, it's needed for coordinators to find only running ```compact``` tasks. At each run, the coordinator computes the available compaction task slots by (total worker capacity - running ```compact``` tasks).

Second, I'm planning to refactor the code around taskStatus. We have several classes relating taskStatus, and meta information (like taskLocation, task creation time, etc) is scattered in those classes. The meta information is important for clients as well as Druid itself, but this scattered information makes difficult to have the same view in all places.

himanshug commented on server/src/main/java/io/druid/client/indexing/ClientCompactQueryTuningConfig.java:
```diff
@@ -0,0 +1,139 @@
+
+public class ClientCompactQueryTuningConfig
+{
+  // These default values should be synchronized with those of IndexTuningConfig
```
I think you meant `RealtimeTuningConfig`. However, in that case, we should probably declare these as `public static final xxx` in one of these classes and other class should refer to those. then they would stay same automatically.

jihoonson replied:
A compaction task is essentially an indexTask + ingestSegmentFirehose. This class should be  synchronized with IndexTuningConfig.
Unfortunately, ```druid-server``` currently has no dependency on ```druid-indexing-service```, so it cannot use the IndexTuningConfig for now. This is why [this discussion](https://github.com/druid-io/druid/pull/5102#discussion_r158430657) comes up.

himanshug replied:
yeah ,  I know ... its unfortunate to have all these classes in `server/` with counterparts in `indexing-service/` . we should probably think about merging them.
maybe the solution is to merge everything from `indexing-service/` into `server/`

created https://github.com/druid-io/druid/issues/5229 to trach it.

himanshug commented on server/src/main/java/io/druid/client/indexing/IndexingWorker.java:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Should be synchronized with io.druid.indexing.worker.Worker
+ */
+public class IndexingWorker
```
is it possible to move specific classes like Worker to server module for now?

himanshug replied:
#5229 would probably be the right solution

himanshug commented on server/src/main/java/io/druid/client/indexing/IndexingWorkerInfo.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Should be synchronized with io.druid.indexing.overlord.ImmutableWorkerInfo
+ */
+public class IndexingWorkerInfo
```
same here

himanshug commented on server/src/main/java/io/druid/client/indexing/QueryStatus.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Should be synchronized with io.druid.indexing.common.TaskStatus.
+ */
+public class QueryStatus
```
same

himanshug commented on server/src/main/java/io/druid/server/coordinator/CoordinatorDynamicConfig.java:
```diff
@@ -69,9 +74,12 @@ public CoordinatorDynamicConfig(
-      @JsonProperty("maxSegmentsInNodeLoadingQueue") int maxSegmentsInNodeLoadingQueue
+      @JsonProperty("killPendingSegmentsSkipList") Object killPendingSegmentsSkipList,
+      @JsonProperty("maxSegmentsInNodeLoadingQueue") int maxSegmentsInNodeLoadingQueue,
+      @JsonProperty("compactionConfigs") List<CoordinatorCompactionConfig> compactionConfigs,
```
this needs to take Object for the same reason as that of killDataSourceWhitelist

jihoonson replied:
Hmm, ```CoordinatorDynamicConfigsResource``` provides an interface to enable/disable compaction for each dataSource like ```POST/GET/DELETE /druid/coordinator/v1/config/compaction/{dataSource}```. Maybe the coordinator console can use this APIs?

himanshug replied:
independent of `/druid/coordinator/v1/config/compaction/{dataSource}` , compaction config can also be updated via general coordinator dynamic config endpoint `POST /druid/coordinator/v1/config`... and coordinator console already has ability to do that.
e.g. after this PR... compaction config would show up in coordinator console and will be update-able there but it would fail because console doesn't have ability to send json list.

jihoonson replied:
Hmm. I'm not much familiar with coffee script and node, but let's fix the coordinator console. Unlike ```killDataSourceWhitelist``` which allows a set of simple strings, ```compactionConfigs``` is a list of ```CoordinatorCompactionConfig``` each of which is a large JSON object. I think it's not a good UI if users should input configuration values as raw JSON objects.

himanshug replied:
My intention wasn't really to change things on UI side and now I see your problem that its not a simple `List<String>`.
However this leads to a problem for users who would use auto compaction, they would not be able to update other coordinate dynamic configs from the console but will have to use manual POST requests for those too.

gianm replied:
My thinking is compaction config should be a different `druid_config` object, separate from the coordinator dynamic config. It doesn't really fit the coordinator dynamic config model which is meant to be a simple form that is filled out on the web console.

The compaction config is more complicated than that so I think it deserves its own config object. At first, it's ok if this is only editable via API.

gianm replied:
@jihoonson, @himanshug what do you think of the above?

jihoonson replied:
@gianm @himanshug ok. I'll make a new config object in a follow-up pr.

gianm replied:
@jihoonson that's ok, but this follow up pr must be done before a release including this feature (to avoid config migration issues) so please raise an issue with milestone 0.13.0 before this current pr is committed.

jihoonson replied:
Raised https://github.com/druid-io/druid/issues/5242.

gianm replied:
Thanks

himanshug replied:
yep ... using separate space for storing compaction config sounds good to me. thanks.

himanshug commented:
looked till `CompactionSegmentIterator` class... will continue later

jihoonson commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,174 @@
+        Map<String, CoordinatorCompactionConfig> compactionConfigs = compactionConfigList
+            .stream()
+            .collect(Collectors.toMap(CoordinatorCompactionConfig::getDataSource, Function.identity()));
+        final List<TaskStatusPlus> runningCompactTasks = indexingServiceClient
```
Changed.

jihoonson commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,174 @@
+        final List<TaskStatusPlus> runningCompactTasks = indexingServiceClient
+            .getRunningTasks()
+            .stream()
+            .filter(status -> status.getType().equals("compact"))
```
TaskStatusPlus guarantees that the type is not null in its constructor. ```CompactionTask.TYPE.equals(status.getType())``` can't be used until #5229.

himanshug replied:
also, I see that "CompactionTask.TYPE" can't be referenced because that is in indexing-service/ .

himanshug replied:
it would be nice to store "compact" as a static final in this class and reference that from CompactionTask.getType() just so that they stay in sync.

jihoonson replied:
Sounds good. Added.

jihoonson commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,174 @@
+        );
+        final int numAvailableCompactionTaskSlots = runningCompactTasks.size() > 0 ?
+                                                    compactionTaskCapacity - runningCompactTasks.size() :
+                                                    Math.max(1, compactionTaskCapacity - runningCompactTasks.size());
```
compactionTaskCapacity might be 0 if totalWorkerCapacity is low. This guarantees that at least one slot is available if compaction is enabled and numRunningCompactTasks is 0.
Added a comment.

himanshug replied:
still not sure why its `Math.max(1, compactionTaskCapacity - numRunningCompactTasks)` and not `Math.max(1, compactionTaskCapacity)` as `numRunningCompactTasks` would be 0 as this point.

jihoonson replied:
Done.

jihoonson commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,174 @@
+        LOG.info("Submit a compactTask[%s] for segments[%s]", taskId, segmentsToCompact);
+
+        if (++numSubmittedCompactionTasks == numAvailableCompactionTaskSlots) {
+          break;
```
Thanks. Fixed.

jihoonson commented on server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorSegmentCompactor.java:
```diff
@@ -0,0 +1,174 @@
+    return stats;
+  }
+
+  public static class SegmentsToCompact
```
Thanks. Moved.

himanshug commented:
:+1: aside from minor comments.

himanshug commented:
forgot to see https://github.com/druid-io/druid/pull/5102#discussion_r160561474 discussion before .
mentioning it here separately in case there are opinions on this.

jihoonson commented:
@himanshug thanks for the review.

himanshug commented:
@jihoonson is this PR rolling upgrade compatible? In many cases, users will upgrade Overlord first and then later Middle Manager . With Overlord running this patch, is it possible that `runningTasks` endpoint doesn't work when Overlord is restarted and taskType is null because MiddleManager announcements stored in zk wouldn't have it.
If that is true, we should probably ignore taskType being null in `runningTasks` endpoint instead of a NPE.

jihoonson commented:
@himanshug thanks for finding it! Raised https://github.com/druid-io/druid/pull/5309.

leventov commented on server/src/main/java/io/druid/server/coordinator/DruidCoordinatorRuntimeParams.java:
```diff
@@ -303,9 +304,30 @@ public Builder withSegmentReplicantLookup(SegmentReplicantLookup lookup)
+          dataSource -> {
+            VersionedIntervalTimeline<String, DataSegment> timeline = dataSources.computeIfAbsent(
+                dataSource.getName(),
+                k -> new VersionedIntervalTimeline<>(String.CASE_INSENSITIVE_ORDER)
```
@jihoonson why `CASE_INSENSITIVE_ORDER`, whereas `DruidCoordinatorCleanupOvershadowed` and `DruidCoordinatorRuleRunner` use natural order for the same thing?

jihoonson replied:
Hmm I don't remember, but it looks wrong. It would be better to call `VersionedIntervalTimeline.forSegments`.

leventov replied:
How do you know that `CASE_INSENSITIVE_ORDER` is wrong? Why natural order is right?

jihoonson replied:
I meant it's wrong in terms of what other places do. But, it shouldn't cause any real bug or error because currently version is a string representation of `DateTime`.
